### DISCUSSION TITLE
Best practice for user activity log

### DISCUSSION THREAD

**Discussion by njxqlus** [filamentphp/filament#10234] [Help]
I'm building a system and I need to store every user action from simple model creation to custom action that calls an external API.

Of course, I know about [Spatie Activity Log Package](https://github.com/spatie/laravel-activitylog) but it has a couple of problems.

### 1. There is no automated logging for mass update

Because [Laravel doesn't fire any useful actions](https://github.com/spatie/laravel-activitylog/issues/240). For example, you can't automatically track bulk delete actions.

### 2. There is no automated logging for sync relationships

Also because [Laravel doesn't fire any useful actions](https://github.com/spatie/laravel-activitylog/issues/679). There is a [altec/eventually package](https://altek.gitlab.io/eventually/installation.html) that can solve it somehow but you need to customize the default behavior of laravel-activitylog.

So I concluded that I couldn't rely on Eloquent Events and any package built on top of them because of there two points above. 
I decided to try to customize the action process of Filament and started from EditResource (update form). There is an option to [customize the save process](https://filamentphp.com/docs/3.x/panels/resources/editing-records#customizing-the-saving-process) but I faced a new problem.

### 3. Relationships using different save processes

For example, if you have a `Select` component based on a relationship in your form you can't control the saving process in the `handleRecordUpdate()` method because the relation has already been saved. Yes, you can use the `saveRelationshipsUsing()` method but this approach brings hell to your code because you must use it on every `Select` and half of your saving logic will be in the `form()` method, and other will be in `EditResource` class. It's also a bad option because you need to customize every plugin like [Spatie Media Library](https://filamentphp.com/plugins/filament-spatie-media-library).

There is an option to use lifecycle hooks but I had the same problem: a relationship saved even before the `beforeSave()` hook.

---

What I want to achieve: after a User pushes a button in Edit Form a record like "User updated field to 123 and set relationship to Foo" will appear in the database.

**Comment by scottzirkel** [comment]
Not really Filament specific, but I've built similar systems in the past. I've created a History trait and added it to whatever models should be tracked. For relationships I've used the 'touches' option from Eloquent. https://laravel.com/docs/10.x/eloquent-relationships#touching-parent-timestamps
